website/docs/r/device_posture_rule.html.markdown

```diff
@@ -69,6 +69,23 @@ The input structure depends on the device posture rule type.
 * `thumbprint` - (Optional) The thumbprint of the application certificate.
 * `running` - (Optional) Checks if the application should be running.
 
+**os_version** allows the following:
+
+* `version` = (Required) The operating system semantic version.
+* `operator` = (Required) The version comparison operator in (>,>=,<,<=,==)
+
+**domain_joined**
+
+* `domain` = (Required) The domain that the client must join.
+
+**firewall**
+
+* `enabled` = (Required) True if the firewall must be enabled.
+
+**disk_encryption**
+
+* `require_all` = (Required) True if all drives must be encrypted.
+
```

Comment:
these descriptions don't match the schema states

```suggestion
* `version` = (Optional) The operating system semantic version.
* `operator` = (Optional) The version comparison operator in (>,>=,<,<=,==)

**domain_joined**

* `domain` = (Optional) The domain that the client must join.

**firewall**

* `enabled` = (Optional) True if the firewall must be enabled.

**disk_encryption**

* `require_all` = (Optional) True if all drives must be encrypted.
```

Reply:
The documentation is correct, the validation is too lax. I'm not sure how to perform validation for the *input* field when it's dependent on the *type* field.

Reply:
you can't really do it; usually the validation goes in the CRUD operations themselves and these fields are marked as optional from the schema point of view.

Reply:
How do you feel about updating the crud operations validation to include something like this?

```
func setDevicePostureRuleInput(rule *cloudflare.DevicePostureRule, d *schema.ResourceData) error {
	if _, ok := d.GetOk("input"); ok {
		input := cloudflare.DevicePostureRuleInput{}

		switch rule.Type {
		case "file":
			p, ok := d.GetOk("input.0.path")
			if !ok {
				return fmt.Errorf("file rule requires input.path")
			}
			input.Path = p.(string)

			exists, ok := d.GetOk("input.0.exists")
			if !ok {
				// File rule exists to true.
				exists = true
			}
			input.Exists = exists.(bool)

			if tp, ok := d.GetOk("input.0.thumbprint"); ok {
				input.Thumbprint = tp.(string)
			}

			if s, ok := d.GetOk("input.0.sha256"); ok {
				input.Sha256 = s.(string)
			}
		case "application":
			p, ok := d.GetOk("input.0.path")
			if !ok {
				return fmt.Errorf("application rule requires input.path")
			}
			input.Path = p.(string)

			if tp, ok := d.GetOk("input.0.thumbprint"); ok {
				input.Thumbprint = tp.(string)
			}

			if s, ok := d.GetOk("input.0.sha256"); ok {
				input.Sha256 = s.(string)
			}

			running, ok := d.GetOk("input.0.running")
			if !ok {
				// Application rule defaults running to true.
				running = true
			}
			input.Running = running.(bool)
		case "serial_number":
			inputID, ok := d.GetOk("input.0.id")
			if !ok {
				return fmt.Errorf("serial_number rule requires input.id")
			}
			input.ID = inputID.(string)
		case "disk_encryption":
			require_all, ok := d.GetOk("input.0.require_all")
			if !ok {
				// Disk encryption rule defaults enabled to true.
				require_all = false
			}
			input.RequireAll = require_all.(bool)
		case "firewall":
			enabled, ok := d.GetOk("input.0.enabled")
			if !ok {
				// Firewall rule defaults enabled to true.
				enabled = true
			}
			input.Enabled = enabled.(bool)
		case "os_version":
			if version, ok := d.GetOk("input.0.version"); ok {
				input.Version = version.(string)
			}
			if operator, ok := d.GetOk("input.0.operator"); ok {
				input.Operator = operator.(string)
			}
		case "domain_joined":
			domain, ok := d.GetOk("input.0.domain")
			if !ok {
				return fmt.Errorf("domain rule requires input.domain")
			}
			input.Domain = domain.(string)
		}

		rule.Input = input
	}

	return nil
}
```

Reply:
we generally use documentation and examples to describe the correct configuration and leave the more complex validation to the API itself. this avoids duplication in multiple systems such as here and `cf-terraforming` and it's very difficult to keep it all in sync correctly.